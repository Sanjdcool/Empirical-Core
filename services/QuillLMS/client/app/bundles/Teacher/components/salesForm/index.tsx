import * as React from 'react';

import LowerFormFields from './lowerFormFields';
import SchoolAndDistrictFields from './schoolAndDistrictFields';
import UpperFormFields from './upperFormFields';

import { getSchoolsAndDistricts, validateSalesForm, submitSalesForm } from '../../helpers/salesForms';
import { InputEvent } from '../../../Staff/interfaces/evidenceInterfaces';
import {
  FIRST_NAME, LAST_NAME, EMAIL, PHONE_NUMBER, ZIPCODE, SCHOOL_PREMIUM_ESTIMATE, TEACHER_PREMIUM_ESTIMATE,
  STUDENT_PREMIUM_ESTIMATE, COMMENTS, SCHOOL, DISTRICT, SCHOOL_OR_DISTRICT, SCHOOL_NOT_LISTED, DISTRICT_NOT_LISTED, SUBMISSION_ERROR
} from '../../../../constants/salesForm';

export const SalesForm = ({ type }) => {
  const [errors, setErrors] = React.useState<any>({});
  const [firstName, setFirstName] = React.useState<string>('');
  const [lastName, setLastName] = React.useState<string>('');
  const [email, setEmail] = React.useState<string>('');
  const [phoneNumber, setPhoneNumber] = React.useState<string>('');
  const [zipcode, setZipcode] = React.useState<string>('');
  const [schoolPremimumEstimate, setSchoolPremiumEstimate] = React.useState<string>('');
  const [teacherPremimumEstimate, setTeacherPremiumEstimate] = React.useState<string>('');
  const [studentPremimumEstimate, setStudentPremiumEstimate] = React.useState<string>('');
  const [comments, setComments] = React.useState<string>('');
  const [schools, setSchools] = React.useState<any[]>([]);
  const [districts, setDistricts] = React.useState<any[]>([]);
  const [selectedSchool, setSelectedSchool] = React.useState<string>('');
  const [schoolNotListed, setSchoolNotListed] = React.useState<boolean>(false);
  const [selectedDistrict, setSelectedDistrict] = React.useState<string>('');
  const [districtNotListed, setDistrictNotListed] = React.useState<boolean>(false);
  const [schoolOrDistrict, setSchoolOrDistrict] = React.useState<any>('');
  const [formSubmitted, setFormSubmitted] = React.useState<boolean>(false);

  React.useEffect(() => {
    if(!schools.length) {
      getSchoolsAndDistricts('school').then((response) => {
        if(response && response.options) {
          setSchools(response.options)
        }
      });
    }
    if(!districts.length) {
      getSchoolsAndDistricts('district').then((response) => {
        if(response && response.options) {
          setDistricts(response.options)
        }
      });
    }
  }, []);

  const stateSetters = {
    [FIRST_NAME]: setFirstName,
    [LAST_NAME]: setLastName,
    [EMAIL]: setEmail,
    [PHONE_NUMBER]: setPhoneNumber,
    [ZIPCODE]: setZipcode,
    [SCHOOL_PREMIUM_ESTIMATE]: setSchoolPremiumEstimate,
    [TEACHER_PREMIUM_ESTIMATE]: setTeacherPremiumEstimate,
    [STUDENT_PREMIUM_ESTIMATE]: setStudentPremiumEstimate,
    [COMMENTS]: setComments,
    [SCHOOL]: setSelectedSchool,
    [DISTRICT]: setSelectedDistrict,
    [SCHOOL_OR_DISTRICT]: setSchoolOrDistrict
  }
  const schoolIsSelected = schoolOrDistrict === SCHOOL;
  const districtIsSelected = schoolOrDistrict === DISTRICT;

  function handleUpdateField(e: InputEvent | React.ChangeEvent<HTMLTextAreaElement>) {
    const { target } = e;
    const { value, id } = target;
    const setterFunction = stateSetters[id];
    if(id === SCHOOL_OR_DISTRICT) {
      // reset to displaying search inputs in case user changes from school to district option (and vice versa)
      setSchoolNotListed(false);
      setDistrictNotListed(false);
    }
    setterFunction(value);
  }

  function handleSchoolSearchChange(value: string) {
    if(value === SCHOOL_NOT_LISTED) {
      setSchoolNotListed(true);
      setSelectedSchool('');
    } else {
      setSelectedSchool(value);
    }
  };

  function handleDistrictSearchChange(value: string) {
    if(value === DISTRICT_NOT_LISTED) {
      setDistrictNotListed(true);
      setSelectedDistrict('');
    } else {
      setSelectedDistrict(value);
    }
  };

  function handleFormSubmission(e: React.MouseEvent) {
    e.preventDefault();
    const salesFormSubmission = {
      first_name: firstName,
      last_name: lastName,
      email: email,
      phone_number: phoneNumber,
      zipcode: zipcode,
      collection_type: schoolOrDistrict.toLowerCase(),
      school_name: selectedSchool,
      district_name: selectedDistrict,
      school_premium_count_estimate: parseInt(schoolPremimumEstimate),
      teacher_premium_count_estimate: parseInt(teacherPremimumEstimate),
      student_premium_count_estimate: parseInt(studentPremimumEstimate),
      submission_type: type,
      comment: comments
    }
    const formErrors = validateSalesForm(salesFormSubmission);
    if(Object.keys(formErrors).length) {
      setErrors(formErrors)
    } else {
      setErrors({});
      submitSalesForm(salesFormSubmission).then(response => {
        if(response.error) {
          const submissionError = { [SUBMISSION_ERROR]: response.error };
          setErrors(submissionError)
        } else {
          setFormSubmitted(true);
        }
      });
    }
  }

  return(
    <div className="sales-form-container">
      {!formSubmitted && <form className="container">
        <UpperFormFields
          email={email}
          errors={errors}
          firstName={firstName}
          handleUpdateField={handleUpdateField}
          lastName={lastName}
          phoneNumber={phoneNumber}
          type={type}
          zipcode={zipcode}
        />
        <SchoolAndDistrictFields
          districtIsSelected={districtIsSelected}
          districtNotListed={districtNotListed}
          districts={districts}
          errors={errors}
          handleDistrictSearchChange={handleDistrictSearchChange}
          handleSchoolSearchChange={handleSchoolSearchChange}
          handleUpdateField={handleUpdateField}
          schoolIsSelected={schoolIsSelected}
          schoolNotListed={schoolNotListed}
          schools={schools}
          selectedDistrict={selectedDistrict}
          selectedSchool={selectedSchool}
        />
        <LowerFormFields
          comments={comments}
          errors={errors}
          handleUpdateField={handleUpdateField}
          schoolPremimumEstimate={schoolPremimumEstimate}
          studentPremimumEstimate={studentPremimumEstimate}
          teacherPremimumEstimate={teacherPremimumEstimate}
        />
        {errors[SUBMISSION_ERROR] && <p className="error-text">{errors[SUBMISSION_ERROR]}</p>}
        <button className="submit-button quill-button contained primary medium" onClick={handleFormSubmission}>Submit</button>
      </form>}
      {formSubmitted && <div className="success-container"><h3>Request successfully submitted!</h3></div>}
    </div>
  )
}

export default SalesForm
