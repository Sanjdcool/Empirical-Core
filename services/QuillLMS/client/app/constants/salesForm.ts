export const FIRST_NAME = 'First name';
export const LAST_NAME = 'Last name';
export const EMAIL = 'Email';
export const PHONE_NUMBER = 'Phone number';
export const ZIPCODE = 'Zipcode';
export const SCHOOL_PREMIUM_ESTIMATE = 'Estimated number of schools that will receive Quill Premium';
export const TEACHER_PREMIUM_ESTIMATE = 'Estimated number of teachers that will receive Quill Premium';
export const STUDENT_PREMIUM_ESTIMATE = 'Estimated number of students that will receive Quill Premium';
export const COMMENTS = 'Comments';
export const SCHOOL = 'School';
export const DISTRICT = 'District';
export const SCHOOL_OR_DISTRICT = 'School or district';
export const SCHOOL_NOT_LISTED = "My school isn’t listed";
export const DISTRICT_NOT_LISTED = "My district isn’t listed";
