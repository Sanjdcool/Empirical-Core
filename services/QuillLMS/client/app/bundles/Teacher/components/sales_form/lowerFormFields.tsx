import * as React from 'react';

import { Input } from '../../../Shared';
import { SCHOOL_PREMIUM_ESTIMATE, TEACHER_PREMIUM_ESTIMATE, STUDENT_PREMIUM_ESTIMATE, COMMENTS } from '../../../../constants/salesForm';

export const LowerFormFields = ({ errors, handleUpdateField, schoolPremimumEstimate, teacherPremimumEstimate, studentPremimumEstimate, comments }) => {
  return(
    <React.Fragment>
      <Input
        className="school-premium-estimate"
        error={errors[SCHOOL_PREMIUM_ESTIMATE]}
        handleChange={handleUpdateField}
        id={SCHOOL_PREMIUM_ESTIMATE}
        label={SCHOOL_PREMIUM_ESTIMATE}
        placeholder=""
        value={schoolPremimumEstimate}
      />
      <Input
        className="teacher-premium-estimate"
        error={errors[TEACHER_PREMIUM_ESTIMATE]}
        handleChange={handleUpdateField}
        id={TEACHER_PREMIUM_ESTIMATE}
        label={TEACHER_PREMIUM_ESTIMATE}
        placeholder=""
        value={teacherPremimumEstimate}
      />
      <Input
        className="student-premium-estimate"
        error={errors[STUDENT_PREMIUM_ESTIMATE]}
        handleChange={handleUpdateField}
        id={STUDENT_PREMIUM_ESTIMATE}
        label={STUDENT_PREMIUM_ESTIMATE}
        placeholder=""
        value={studentPremimumEstimate}
      />
      <div className="control">
        <label className="label" htmlFor={COMMENTS} id="comments-label">Comments (optional)</label>
        <textarea
          aria-labelledby="comments-label"
          id={COMMENTS}
          onChange={handleUpdateField}
          style={{minHeight: '100px', border: '1px solid black', padding: '10px', width: '100%'}}
          value={comments}
        />
      </div>
    </React.Fragment>
  );
}

export default LowerFormFields;
