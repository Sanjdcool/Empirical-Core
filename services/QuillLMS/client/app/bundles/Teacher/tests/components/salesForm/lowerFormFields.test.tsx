import * as React from 'react';
import { mount } from 'enzyme';

import { LowerFormFields }from '../../../components/salesForm/lowerFormFields';

jest.mock('fuse.js', () => ({
  default: jest.fn()
}));

describe('LowerFormFields Component', () => {
  const mockProps = {
    errors: {},
    handleUpdateField: jest.fn(),
    schoolPremimumEstimate: 1,
    teacherPremimumEstimate: 20,
    studentPremimumEstimate: 400,
    comments: ''
  }
  const component = mount(<LowerFormFields {...mockProps} />);

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot();
  });
});
