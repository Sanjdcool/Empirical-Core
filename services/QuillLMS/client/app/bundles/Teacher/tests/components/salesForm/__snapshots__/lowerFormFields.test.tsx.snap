// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LowerFormFields Component should match snapshot 1`] = `
<LowerFormFields
  comments=""
  errors={Object {}}
  handleUpdateField={[MockFunction]}
  schoolPremimumEstimate={1}
  studentPremimumEstimate={400}
  teacherPremimumEstimate={20}
>
  <Input
    autoComplete="on"
    className="form-input estimate"
    handleChange={[MockFunction]}
    id="Estimated number of schools that will receive Quill Premium"
    label="Estimated number of schools that will receive Quill Premium"
    placeholder=""
    value={1}
  >
    <div
      className="input-container  inactive has-text form-input estimate "
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <label
        htmlFor="Estimated number of schools that will receive Quill Premium"
      >
        Estimated number of schools that will receive Quill Premium
      </label>
      <input
        autoComplete="on"
        id="Estimated number of schools that will receive Quill Premium"
        maxLength={10000}
        onChange={[MockFunction]}
        onFocus={[Function]}
        placeholder={false}
        value={1}
      />
    </div>
  </Input>
  <Input
    autoComplete="on"
    className="form-input estimate"
    handleChange={[MockFunction]}
    id="Estimated number of teachers that will receive Quill Premium"
    label="Estimated number of teachers that will receive Quill Premium"
    placeholder=""
    value={20}
  >
    <div
      className="input-container  inactive has-text form-input estimate "
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <label
        htmlFor="Estimated number of teachers that will receive Quill Premium"
      >
        Estimated number of teachers that will receive Quill Premium
      </label>
      <input
        autoComplete="on"
        id="Estimated number of teachers that will receive Quill Premium"
        maxLength={10000}
        onChange={[MockFunction]}
        onFocus={[Function]}
        placeholder={false}
        value={20}
      />
    </div>
  </Input>
  <Input
    autoComplete="on"
    className="form-input estimate"
    handleChange={[MockFunction]}
    id="Estimated number of students that will receive Quill Premium"
    label="Estimated number of students that will receive Quill Premium"
    placeholder=""
    value={400}
  >
    <div
      className="input-container  inactive has-text form-input estimate "
      onClick={[Function]}
      onKeyDown={[Function]}
      role="button"
      tabIndex={-1}
    >
      <label
        htmlFor="Estimated number of students that will receive Quill Premium"
      >
        Estimated number of students that will receive Quill Premium
      </label>
      <input
        autoComplete="on"
        id="Estimated number of students that will receive Quill Premium"
        maxLength={10000}
        onChange={[MockFunction]}
        onFocus={[Function]}
        placeholder={false}
        value={400}
      />
    </div>
  </Input>
  <div
    className="comments-container"
  >
    <label
      htmlFor="Comments"
      id="comments-label"
    >
      Comments (optional)
    </label>
    <textarea
      aria-labelledby="comments-label"
      id="Comments"
      onChange={[MockFunction]}
      style={
        Object {
          "border": "1px solid black",
          "minHeight": "100px",
          "padding": "10px",
          "width": "100%",
        }
      }
      value=""
    />
  </div>
</LowerFormFields>
`;
